### PR TITLE
allow annotation on extern methods

### PR DESCRIPTION
**Pull request by hanw**

### CODE REVIEW

File: frontends/parsers/p4/p4parser.ypp [hunk 1]
```diff
@@ -556,12 +556,13 @@ functionPrototype
...
-    : functionPrototype ";" { $$ = $1; }
-    | ABSTRACT functionPrototype ";"    { $$ = $2; $$->setAbstract(); }  // experimental
-    | TYPE "(" parameterList ")" ";"  // constructor
-                                        { auto par = new IR::ParameterList(@3, *$3);
-                                          auto mt = new IR::Type_Method(@1, par);
-                                          $$ = new IR::Method(@1, IR::ID(@1, $1), mt); }
+    : optAnnotations functionPrototype ";" { $$ = $2; $2->annotations = $1; }
+    | optAnnotations ABSTRACT functionPrototype ";"    { $$ = $3; $$->setAbstract();
+                                                         $3->annotations = $1; }  // experimental
+    | optAnnotations TYPE "(" parameterList ")" ";"  // constructor
```
Comment (mbudiu-vmw) [inline frontends/parsers/p4/p4parser.ypp]: I don't know if annotations on the constructor are useful; you already have annotations on the type declaration.

Comment (mbudiu-vmw) [reply frontends/parsers/p4/p4parser.ypp]: In fact, you are not using the annotations here.

Comment (hanw) [reply frontends/parsers/p4/p4parser.ypp]: Annotation on constructor is used in the 4th argument in `new IR::Method()`

Comment (hanw) [reply frontends/parsers/p4/p4parser.ypp]: Without the `optAnnotations` on the constructor, there is a shift/reduce conflict.

**Comment by hanw** [comment]
I tested with the following program, and `dump` the top level `IR::P4Program`. How do I add a test case with reference output to the test suite?

```
enum counter_type_t {
    BYTES,
    PACKETS
}
@annotate("test")
extern counter<I> {
    @annotate("test")
    counter(I inst);
    @annotate("test")
    void count(counter_type_t type, in I count);
}
```

**Comment by mbudiu-vmw** [comment]
To add a test copy the file to the testsuite/p4_16_samples and rerun 'cmake ..; make check'.
This will generate all reference outputs, which you can then add to git.

**Comment by hanw** [comment]
@mbudiu-vmw PR for spec https://github.com/p4lang/p4-spec/pull/439

File: testdata/p4_16_samples_outputs/annotations-first.p4 [hunk 2]
```diff
@@ -1,6 +1,7 @@
 @annotest const bit<1> b = 1w1;
 extern Annotated {
     Annotated();
+    void execute(bit<8> index);
```
Comment (ChrisDodd) [inline testdata/p4_16_samples_outputs/annotations-first.p4]: looks like ToP4 is failing to output the annotations on the methods?  Or is something deleting them?

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/annotations-first.p4]: In this case you may have to do some more work in toP4.cpp

**Comment by mbudiu-vmw** [comment]
This is not finished, you have to implement the support in toP4, as @ChrisDodd pointed.

**Comment by hanw** [comment]
addressed Chris's comment in #967

**Pull request merged**